test/e2e/ingress/disable_merge_slashes_test.go

```diff
@@ -0,0 +1,99 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build e2e
+// +build e2e
+
+package ingress
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo/v2"
+	"github.com/projectcontour/contour/test/e2e"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func testIngressDisableMergeSlashes(disableMergeSlashes bool) e2e.NamespacedTestBody {
+	var testName string
+	if disableMergeSlashes {
+		testName = "multiple slashes in path should not be merged"
+	} else {
+		testName = "multiple slashes in path should be merged"
+	}
+	return func(namespace string) {
+		Specify(testName, func() {
+			t := f.T()
+
+			f.Fixtures.Echo.Deploy(namespace, "echo")
+
+			var hostName string
+			if disableMergeSlashes {
+				hostName = "disable-merge-slashes.projectcontour.io"
+			} else {
+				hostName = "merge-slashes.projectcontour.io"
+			}
+
+			i := &networkingv1.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: namespace,
+					Name:      "ingress-with-disable-merge-slashes",
+				},
+				Spec: networkingv1.IngressSpec{
+					Rules: []networkingv1.IngressRule{
+						{
+							Host: hostName,
+							IngressRuleValue: networkingv1.IngressRuleValue{
+								HTTP: &networkingv1.HTTPIngressRuleValue{
+									Paths: []networkingv1.HTTPIngressPath{
+										{
+											PathType: e2e.IngressPathTypePtr(networkingv1.PathTypePrefix),
+											Path:     "/",
+											Backend: networkingv1.IngressBackend{
+												Service: &networkingv1.IngressServiceBackend{
+													Name: "echo",
+													Port: networkingv1.ServiceBackendPort{
+														Number: 80,
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+
+			require.NoError(t, f.Client.Create(context.TODO(), i))
+
+			res, ok := f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+				Host:      i.Spec.Rules[0].Host,
+				Path:      "/path/with//multiple/slashes",
+				Condition: e2e.HasStatusCode(200),
+			})
+			require.NotNil(t, res, "request never succeeded")
+			require.Truef(t, ok, "expected 200 response code, got %d", res.StatusCode)
+
+			if disableMergeSlashes {
+				assert.Equal(t, "/path/with//multiple/slashes", f.GetEchoResponseBody(res.Body).Path)
+			} else {
+				assert.Equal(t, "/path/with/multiple/slashes", f.GetEchoResponseBody(res.Body).Path)
+			}
```

Comment:
Something like this seems to get rid of the test flake for me (unfortunately `f.HTTP.RequestUntil` doesn't allow you to check conditions on the response body, something to fix but don't need to do it this PR, hence why it's wrapped in a `require.Eventually`).

I'm wondering if sometimes the Route update gets to Envoy before the Listener update, so the request returns a 200 but the listener hasn't actually been reconfigured yet? @sunjayBhatia does that sound possible to you?

```suggestion
			require.Eventually(t, func() bool {
				res, ok := f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
					Host:      i.Spec.Rules[0].Host,
					Path:      "/path/with//multiple/slashes",
					Condition: e2e.HasStatusCode(200),
				})
				if !ok || res == nil {
					return false
				}

				if disableMergeSlashes {
					return "/path/with//multiple/slashes" == f.GetEchoResponseBody(res.Body).Path
				} else {
					return "/path/with/multiple/slashes" == f.GetEchoResponseBody(res.Body).Path
				}
			}, time.Minute, time.Second)
```

Reply:
yeah, since the Listener resource stream and Route resource streams are separate, you can get into that case (which is an other advertisement for using the ADS version of xDS so things are ordered properly for us)

might be able to fix it by making this a TLS enabled Ingress so requests have to go through a different filter chain to be accepted, so the listener update has to happen before a request is successful

Reply:
FYI we've merged https://github.com/projectcontour/contour/pull/4409, so you could now implement the change above directly within the `Condition` function in the `f.HTTP.RequestUntil` call and drop the `require.Eventually`, or @sunjayBhatia's suggestion is another viable path.